duckduckgo/Android
# Improve URL handling so we display the correct url at the correct time

Task/Issue URL: 
- https://app.asana.com/0/361428290920652/1125745835404857
- https://app.asana.com/0/361428290920652/1129928100535218
- https://app.asana.com/0/361428290920652/1132109259549317

Tech Design URL: https://app.asana.com/0/361428290920652/1125745835404855

**Description**:
- Improve URL handling so we display the correct url at the correct time, this improves user experience and protect against phishing attacks.
- Also fix an issue where omnibar text from previous url could show on home screen
- Make extraction of website title hardier so that it is displayed correctly in tabs

**Steps to test this PR**:

**Check that url, privacy dash and back/forward buttons update as expected during navigation:**
1. Visit facebook.com and note the url updates in the omnibar and the privacy dash is refreshed
1. Visit reddit.com and note the url updates in the omnibar and the privacy dash is refreshed
1. Tap a subreddit link and note the url updates again (this previously did not happen)
1. Now press back all the way to the home screen to make sure navigation works as expected
1. Ensure the omnibar displays no url when on home
1. Now use the forward multiple times to return to the subreddit page again

**Check against historical phishing attack:**
1. Visit https://soda.fm/spoof1.html
1. Ensure that the url does not change until off the spoofed page

**Check against latest phishing attack:**
1. Visit https://geekboy.ninja/p0c/spoof-bar.html
1. Ensure that the url does not change

**Check refresh works as it did historically**
1. Visit cnn.com
1. Note the number of trackers in the privacy dashboard
1. Select refresh to refresh the page and open the privacy dashboard again
1. Note that the number of trackers resets and repopulates
NOTE: Refresh relies on onPageStarted as it used to. Fixing this was beyond the scope of this project however I have created https://app.asana.com/0/361428290920652/1131849703601622 to cover it at a later date

**Check we only make fire one atb request when navigating to our serp**
1. Open Charles to watch app traffic
1. Perform a search to visit our serp
1. Ensure that the atb call only fires once

**Check other urls**
- Tabs
- Bookmarks
- Sharing
- Pin to home

---
###### Internal references:
[Software Engineering Expectations](https://app.asana.com/0/59792373528535/199064865822552)
[Technical Design Template](https://app.asana.com/0/59792373528535/184709971311943)
